Docs 7582 invalidate cache v4.3 cv (#1613)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/cache-scope-invalidate.adoc

= Invalidate Cache Entries
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> You can invalidate a cache key or a complete cache by using the Invalidate Key and Invalidate Cache components respectively. Alternatively, you can configure the caching strategy to invalidate entries automatically based on time.

Reviewer: ```suggestion
You can invalidate a cache key or a complete cache by using the Invalidate Key and Invalidate Cache components, respectively. Alternatively, you can configure the caching strategy to invalidate entries automatically based on time.
```
Do we need to say more about "based on time"; are we saying that they time out based on a setting you make?

> Invalidate cache entries when you want the Cache Scope to recalculate the value for those cache entries.

Reviewer: I think it might be good to state the use case first: what is the user goal, under what circumstances does the suer want to recalculate the value of the cache entries? And then tell them that they can do that, using these components.

== Invalidate a Cache Key

Use the Invalidate Key component to invalidate the specified cache key in the referenced caching strategy.

=== Configure the Invalidate Key Component

> To invalidate a cache key:

Reviewer: The container topic is Invalidate a Cache Key, which contains Configure the IKC and Example IK Config.
So...maybe the intro text should mention that to invalidate, you need to configure the IKC first?

> To invalidate a cache key:

Reviewer: This is the sentence that needs adjustment, because it looks like it belongs under the container rather than here:
```suggestion
== Invalidate a Cache Key

Use the Invalidate Key component to invalidate the specified cache key in the referenced caching strategy.

=== Configure the Invalidate Key Component

To invalidate a cache key:
```
(no changes suggested, for you to evaluate....)

. In *Anypoint Studio*, drag the *Invalidate Key* component to your flow.

> . Open the *Invalidate Key* properties.

Reviewer: How does the user do this: "open"? Click? Select? Expand?

> . Specify the *Caching Strategy* where the cache key to invalidate is stored.

Reviewer: Is this what we mean:
```suggestion
. In *Caching Strategy*, specify where the cache key to invalidate is stored.
```
Does specifying this actually result in invalidation?
I guess I am struggling with the fact that invalidation is performed by...reconfiguration?

> . Optionally, select the method for generating the cache key.

Reviewer: "generating"? but I thought we were invalidating it? Is this a replacement key?

> === Example Invalidate Key Configuration

Reviewer: ```suggestion
=== Example Invalidate Key Configuration
```
or
```suggestion
=== Example Invalidate Key Component Configuration
```

[source, xml, linenums]
----
<!-- Caching strategy configuration -->
<ee:object-store-caching-strategy name="myCachingStrategy" keyGenerationExpression="#[payload.key]" />

<flow name="cacheFlow">
  <ee:cache cachingStrategy-ref="myCachingStrategy">
      <!-- Processing logic inside the Cache Scope -->
  </ee:cache>
</flow>

<flow name="invalidateItemFlow">
  <!-- Invalidate Key component configuration -->
  <ee:invalidate-key cachingStrategy-ref="myCachingStrategy" keyGenerationExpression="#[payload.key]" />
</flow>
----

== Invalidate a Complete Cache

Use the Invalidate Cache component to invalidate the complete cache in the referenced caching strategy.

=== Configure the Invalidate Cache Component

> To invalidate a complete cache:

Reviewer: Same wording confusion (on my part) as in the previous....

. In *Anypoint Studio*, drag the *Invalidate Cache* component to your flow.

> . Open the *Invalidate Cache* properties.

Reviewer: ditto

. Specify the *Caching Strategy* to invalidate.

=== Example Invalidate Cache Configuration

[source, xml, linenums]
----
<!-- Caching strategy configuration -->
<ee:object-store-caching-strategy name="myCachingStrategy" />

<flow name="cacheFlow">
  <ee:cache cachingStrategy-ref="myCachingStrategy">
      <!-- Processing logic inside the Cache Scope -->
  </ee:cache>
</flow>

<flow name="invalidateCompleteCacheFlow">
  <!-- Invalidate Cache component configuration -->
  <ee:invalidate-cache cachingStrategy-ref="myCachingStrategy" />
</flow>
----

== Automatic Invalidation

When you configure the caching strategy, you can configure a custom object store and set the time to live (`entryTtl`) value to make stored cache keys expire over time.

> === Configure the Object Store in your Caching Strategy

Reviewer: ```suggestion
=== Configure the Object Store in Your Caching Strategy
```

. Follow the steps in xref:cache-scope-strategy.adoc[Set Up a Caching Strategy]
. When configuring the object store for the strategy, set a value for *Entry ttl*.
+
This value represents the time after which the cache entry expires.

> . Complete all the necessary values for the object store configuration.

Reviewer: ```suggestion
. Finish providing all the necessary values for the object store configuration.
```
or
```suggestion
. Complete the object store configuration.
```

> === Example Caching Strategy Configuration

Reviewer: Isn't this an example configuration of automatic invalidation?

Reviewer: Yeah, so for this topic reexamine the title hierarchy and whether it makes clear how the subtopics relate to the container topics. 
I wonder if it would be more clear with titles of Steps to Configure... and Example Configuration of...?
= Invalidate Cache Entries
== Invalidate a Cache Key
=== Configure the Invalidate Key Component 
=== Example Invalidate [Cache?] Key Configuration
== Invalidate a Complete Cache
=== Configure the Invalidate Cache Component
=== Example Invalidate [Complete?] Cache Configuration
== Automatic Invalidation
=== Configure the Object Store in Your Caching Strategy
=== Example Caching Strategy Configuration [Example Automatic Invalidation Configuration?]

[source, xml, linenums]
----
<!-- Object Store configuration-->
<os:config name="ObjectStore_Config"/>
<!-- Caching Strategy configuration-->
<ee:object-store-caching-strategy name="Caching_Strategy">
  <!-- Object Store defined for the caching strategy -->
  <os:private-object-store
    alias="CachingStrategy_ObjectStore"
    maxEntries="100"
    entryTtl="10"
    expirationInterval="5"
    config-ref="ObjectStore_Config"/>
</ee:object-store-caching-strategy>
----

== See Also

* xref:cache-scope-to-configure.adoc[Configure the Cache Scope]
* xref:cache-scope-strategy.adoc[Set Up a Caching Strategy]

> * xref:mule-object-stores.adoc#object-store-config[Object Store Configuration]

Reviewer: I wonder if the invalidation topics belong in these containers rather than off on their own?

= Invalidate Cache Entries
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

You can invalidate a cache key or a complete cache by using the Invalidate Key and Invalidate Cache components respectively.

> Invalidate cache entries when you want the Cache Scope to execute all processors to generate a response, instead of returning a cached response for the existing entry.

Reviewer 2: i would rather express this as `want the Cache Scope to recalculate the value for a given cache entry`

== Invalidate a Cache Key

Use the Invalidate Key component to invalidate the specified cache key in the referenced caching strategy.

=== Configure the Invalidate Key Component

To invalidate a cache key:

. In *Anypoint Studio*, drag the *Invalidate Key* component to your flow.
. Open the *Invalidate Key* properties.
. Specify the *Caching Strategy* where the cache key to invalidate is stored.

> . Select the method for generating the cache key to search and invalidate.

Reviewer 2: not sure what this is intended to mean.

> Ensure that the method selected for generating the cache key in the Invalidate Component is the same used in the referenced Caching Strategy, otherwise the cache keys might not match and the process won't work as expected.

Reviewer 2: this would assume that the payload of the flow using the cache and the flow invalidating it are the same, which is not necessarily the case.

=== Example Invalidate Key Configuration

[source, xml, linenums]
----
<!-- Caching strategy configuration -->
<ee:object-store-caching-strategy name="myCachingStrategy" keyGenerationExpression="#[payload.key]" />

<flow name="cacheFlow">
  <ee:cache cachingStrategy-ref="myCachingStrategy">

>       <!-- Processing logic to be cached -->

Reviewer 2: be careful with the wording here. What is cached is the final result of this chain, not the logic
